API functions for getting active vs. console vs. source editor

kevinushey opened the pull request:
NOTE: Not ready to merge -- pending other work that needs to be done in tandem with this.

Intended to help resolve https://github.com/rstudio/addinexamples/issues/2.

This PR adds three new API functions for getting an editor context:

- `getLastActiveEditorContext()`,
- `getConsoleEditorContext()`,
- `getSourceEditorContext()`.

It also ensures that `getLastActiveEditorContext()` handles the case where there is no source document, but the console does not have focus.

## TODO
- [ ] Update `rstudioapi` accordingly
- [x] Ensure that we're backwards-compatible with old API
- [ ] Update `addinexamples` accordingly

jjallaire commented:
We should definitely try to merge this right away if we can (to make the release window). Can we merge now and update the other repos after?

kevinushey commented:
I think this should be safe to merge -- it should preserve existing functionality; we just have the new endpoints available now for when we update `rstudioapi` and thus `addinexamples`.

jjallaire commented:
@jmcphers Could you also review this? It's a big diff but most of that is just renaming. @kevinushey If you could highlight any pieces that you think bear special scrutiny that would be helpful.

BTW our last rstudioapi submission was actually Dec 9th so we can submit a new version anytime. I'd be in favor of doing this soon so that addins can take advantage of the new APIs

kevinushey commented on src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:
```diff
@@ -731,6 +708,15 @@ else if (uiPrefs_.enableEmacsKeybindings().getGlobalValue())
       // handle chunk options event
       handleChunkOptionsEvent();
+      
+      consoleEditorProvider_.getConsoleEditor().addFocusHandler(new FocusHandler()
```
Do we have any guarantee that the `Console` is constructed eagerly enough (ie, before we construct `Source`) that we won't run into race conditions here? Note that the `consoleEditorProvider_` doesn't get initialized until here: https://github.com/rstudio/rstudio/blob/master/src/gwt/src/org/rstudio/studio/client/workbench/views/console/shell/Shell.java#L114

kevinushey commented on src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:
```diff
@@ -3132,6 +3141,7 @@ public void onSelection(SelectionEvent<Integer> event)
       if (event.getSelectedItem() >= 0)
       {
          activeEditor_ = editors_.get(event.getSelectedItem());
+         consoleEditorHadFocusLast_ = false;
```
This was the 'best' spot I could find for hooking into changes in the active editor. It's a bit asymmetric, but noting that:

1. We listen for console focus events -- when the console editor is focused, `consoleEditorHadFocusLast_` is set to `true`;
2. When a source editor is selected, we run through this bit of code (setting the `activeEditor_` in the process), and also set `consoleEditorHadFocusLast_` to `false`.

It feels a bit unnatural but it seems to work.

jjallaire replied:
I'm wondering if we think it's okay to let these changes wait until
v0.9.1000 -- are the current APIs problematic for a significant class of
plugins?


On Fri, Jan 22, 2016 at 9:09 PM, Kevin Ushey <notifications@github.com>
wrote:

> In
> src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java
> <https://github.com/rstudio/rstudio/pull/558#discussion_r50613605>:
>
> > @@ -3132,6 +3141,7 @@ public void onSelection(SelectionEvent<Integer> event)
> >        if (event.getSelectedItem() >= 0)
> >        {
> >           activeEditor_ = editors_.get(event.getSelectedItem());
> > +         consoleEditorHadFocusLast_ = false;
>
> This was the 'best' spot I could find for hooking into changes in the
> active editor. It's a bit asymmetric, but noting that:
>
>    1. We listen for console focus events -- when the console editor is
>    focused, consoleEditorHadFocusLast_ is set to true;
>    2. When a source editor is selected, we run through this bit of code
>    (setting the activeEditor_ in the process), and also set
>    consoleEditorHadFocusLast_ to false.
>
> It feels a bit unnatural but it seems to work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/558/files#r50613605>.
>

kevinushey replied:
I think it's fine to leave it. The issue @wch pointed out occurs only when you attempt to use an interactive addin when there is no source document available (ie, only the console is open, and the source window is hidden)

jjallaire replied:
Okay, I'll hold merging this until after we branch then.

On Fri, Jan 22, 2016 at 10:38 PM, Kevin Ushey <notifications@github.com>
wrote:

> In
> src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java
> <https://github.com/rstudio/rstudio/pull/558#discussion_r50615151>:
>
> > @@ -3132,6 +3141,7 @@ public void onSelection(SelectionEvent<Integer> event)
> >        if (event.getSelectedItem() >= 0)
> >        {
> >           activeEditor_ = editors_.get(event.getSelectedItem());
> > +         consoleEditorHadFocusLast_ = false;
>
> I think it's fine to leave it. The issue @wch <https://github.com/wch>
> pointed out occurs only when you attempt to use an interactive addin when
> there is no source document available (ie, only the console is open, and
> the source window is hidden)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/558/files#r50615151>.
>

kevinushey commented:
This should be good to go, although we might want to talk in person about the implementation and convince ourselves that this is the best way to go.

kevinushey commented:
Alright, I'm a lot happier with this PR now. A singleton object is used to track the last focused editor, and all AceEditor instances let that tracker know when they've been focused. This makes it quite easy to introspect as to whether a particular editor belongs to a source pane, or a console pane, and so on.

jmcphers commented on src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorFocusTracker.java:
```diff
@@ -0,0 +1,37 @@
+/*
+ * AceEditor.java
```
nit: copy/paste

kevinushey replied:
Doh, thanks. (really need to just figure out an Eclipse setting that can auto-generate these)

jmcphers commented on src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java:
```diff
@@ -723,15 +726,16 @@ else if (uiPrefs_.enableEmacsKeybindings().getGlobalValue())
+   
+   private boolean consoleEditorHadFocusLast()
+   {
+      AceEditor editor = editorFocusTracker_.getLastFocusedEditor();
```
Does this need to work across source windows? (If so, look at `DocFocusedEvent` and `SourceWindowManager.getLastFocusedSourceWindowId`)

kevinushey replied:
It does -- I'll take a look here. Thanks!

jmcphers commented on src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorFocusTracker.java:
```diff
@@ -0,0 +1,37 @@
+import com.google.inject.Singleton;
+
+@Singleton
+public class AceEditorFocusTracker
```
Could also just do this with a static variable in `Source` (up to you though)

kevinushey replied:
I'm imagining that this might be used across more modules than just `Source` in the future so I think it's okay to have it separate.

jmcphers replied:
Fair enough!

kevinushey commented:
It's becoming quite unwieldy to add a new editor-related addin command, so I think before merging I'll take some time to consolidate this into a single stringly-typed event (just because this can reduce boilerplate / make it easier to add future commands).

jjallaire commented:
Okay, sounds good!

On Wed, Mar 23, 2016 at 8:57 PM, Kevin Ushey <notifications@github.com>
wrote:

> It's becoming quite unwieldy to add a new editor-related addin command, so
> I think before merging I'll take some time to consolidate this into a
> single stringly-typed event (just because this can reduce boilerplate /
> make it easier to add future commands).
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/558#issuecomment-200601783>
>

kevinushey commented:
Okay, I've refactored things quite a bit to make things more easily extensible:

1) Editor commands now have their own single event entrypoint, `editor_command`. We funnel any kinds of events that might need to be targeted at a single editor through this.

2) An `EditorCommandEvent` class handles the actual dispatching to the correct source window, and

3) The `SourceWindowManager` takes care of extracting and firing the event to the appropriate listeners.

kevinushey commented:
@jmcphers, can you review the `AceEditorFocusTracker` and make sure that I'm not doing something that might cause problems? This is basically a singleton provider for an object that lives on the main window, that keeps track of `AceEditor` instances across windows. (note that these are GWT, not pure JavaScript, objects)

jmcphers commented:
I think that this pattern creates a problem. `wnd.$lastFocusedEditor` will give you back something that you may be able to use as an AceEditor in some contexts, but not others (because it's not an instance of an AceEditor in the current window). For instance, checked casts are going to fail, enums aren't going to match, etc. None of that is happening now but I don't think we should make it easy for it to occur.

What do you envision the other usages of the focus tracker to be? How about storing something that uniquely identifies the editor (such as its doc ID, with a special ID for the console) rather than the editor object itself?

kevinushey commented:
Got it -- I agree then that we should just be storing an ID. At this point, we only use it to identify if the last focused editor was the console editor, so we don't need more than that.

kevinushey commented:
@jmcphers I've updated the PR with some extra tools.

First, I've added the `MainWindowObject` class is a helper for getting / setting properties on the main window (well, as part of the property bag `$RStudio` within the main window), and this can be used for ad-hoc communication between windows.

I'm also now using this + a focus handler on the window to track the last focused (source) window -- previously, this was done with an `onDocFocused()` event, but this failed in the following scenario:

1) Focus is within a satellite (source) window,
2) You click on the main window console,

as no `DocFocusEvent` would be fired as a result of 2).

Does this all seem sensible?

jmcphers commented on src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditor.java:
```diff
@@ -263,6 +264,7 @@ public void onLoaded()
    @Inject
    public AceEditor()
    {
+      id_ = StringUtil.makeRandomId(16);
```
Where is this used?

kevinushey commented:
Reviewed with @jmcphers and incorporated necessary changes.

jjallaire commented:
Are there some corresponding changes to rstudioapi to take advantage of the new facilities?

kevinushey commented:
Yep -- I made those changes here. https://github.com/rstudio/rstudioapi/commit/291029c2835b0e6751c95f2d75730d1c35465587

jjallaire commented:
Awesome, thanks!


On Fri, Mar 25, 2016 at 9:21 PM, Kevin Ushey <notifications@github.com>
wrote:

> Yep -- I made those changes here. rstudio/rstudioapi@291029c
> <https://github.com/rstudio/rstudioapi/commit/291029c2835b0e6751c95f2d75730d1c35465587>
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/558#issuecomment-201645354>
>
